### ISSUE TITLE
null values override

### ISSUE THREAD

**Issue by messiahUA** [isometry/terraform-provider-deepmerge#39]
I'm not sure this is an issue, but I've faced a situation when I have a defined variable for example:

```
variable "policies" {
  type = map(map(object({
    name         = optional(string)
    policy_arn   = optional(string)
    })
  ))
}
```
And I want to merge this with a local map that has a name set, for example:
```
policies = {
  group1 = {
    policy1 = {
      name = "something"
    }
  }
}
```

In the input variable, I have policy_arn set, for example:
```
policies = {
  group1 = {
    policy1 = {
      policy_arn = "arn"
    }
  }
}
```

The problem is that the resulting structure will have a name set to null:
```
policies = {
  group1 = {
    policy1 = {
      name        = null
      policy_arn = "arn"
    }
  }
}
```
This is because the type is defined and terraform sets this value to null and merge overwrites the value with null.

So the question is maybe it is worth skipping any null values during merging?

**Comment by isometry** [comment]
Thanks for the report! I agree that the behaviour you describe is suboptimal. I'll try to take a look, else I'd welcome a PR :-)

**Comment by messiahUA** [comment]
Additional thoughts on this matter: I think that it makes sense to nullify the value if null is provided especially considering that for terraform "null" is a value. The fundamental issue here is that terraform provides a "nullable" switch only for the variable as a whole (https://developer.hashicorp.com/terraform/language/values/variables#disallowing-null-input-values) and there is no way to specify this for nested structures, the only way is to use "any" type but this is sub-optimal. So I would say probably the best option is to add a switch for the deepmerge to ignore nulls.

**Comment by isometry** [comment]
My interpretation of your initial report was that a non-null value "on the right" should override an existing null "on the left", which is generally the behaviour that I would expect.

On the other hand, the logic with regards a null "on the right" replacing an existing non-null value "on the left" is far less obvious. I see legitimate use-cases for "keeping the left", "nulling", or even for it being a way to completely remove the nulled key from the final output map. Anything other than the current default behaviour (which, I think should be "keep the left value"?) should definitely be controlled by an option.

**Comment by messiahUA** [comment]
> My interpretation of your initial report was that a non-null value "on the right" should override an existing null "on the left", which is generally the behaviour that I would expect.

Yes, although I would rephrase this simply as "null should not override non-null value", but only when the switch is on because it makes sense to override in other circumstances.

I don't think there is anything else to it and no other functionality is needed.

Switch on:
null -> value = value

Switch off:
null -> value = null

This is needed simply because there is no way to tell Terraform that I don't want "null" for an unset input variable. Otherwise, everything would be correct and simple:

unset -> value = value

**Comment by isometry** [comment]
Having spent some time digging into this a little deeper at the weekend, I'm sad to report that this appears to be rather difficult to achieve without essentially reimplementing `mergo`. If you or anyone else can contribute either an alternate merge implementation (via another function name) or a light patch to the existing `mergo` implementation, then I will be happy to review and accept. I had initially hoped that `mergo`s Transform interface would facilitate this, but unless I'm missing something, they only enable custom merge logic for top-level data structures passed to `mergo`, and are ignored for values found nested within top-level maps.

**Comment by shaneramey** [comment]
A workaround is to use 
```
variable "policies" {
  type = map(any)
}
```
